server/aesdsocket.c

```diff
@@ -0,0 +1,346 @@
+	if(rx_buf==NULL)
+	{
+		syslog(LOG_ERR,"malloc for rx buffer failed\n");
+	}
+	
+	
+	//open in append or  creating file /var/tmp/aesdsocketdata file if it doesn’t exist.
+	outputfile_fd = open(FILE_PATH, O_RDWR | O_APPEND | O_CREAT, 0744);
+	perror("open ouputfile");
+	if (outputfile_fd == -1)
+	{
+		syslog(LOG_ERR, "Error in opening file");
+		free(rx_buf);
+		//free(tx_buf);
+		close_all();
+	}
+	
+	/*-------------------------------------------------------------------*/	
+	int total_rxbytes = 0; //total rx bytes 
+	int buffer_realloc_count = 0; 
+	while(1)
+	{
+		/* accept a connection on a socket, creates a new connected socket, and returns a new file descrtor */
+		client_fd = accept(socket_fd, (struct sockaddr *)&clientsockaddr, &addr_size);
+		if(client_fd == -1)
+		{
+			syslog(LOG_ERR, "Error accepting connection from client\n");
+			return -1;
+		}
+		
+		/* print accepted client IP address */
+		char *client_ip_address = inet_ntoa(clientsockaddr.sin_addr);		
+		syslog(LOG_DEBUG,"Accepted connection from %s", client_ip_address);
+		int current_rxbytes = 0;
+		
+		
+		/*recieve all data in buffer by realllcating buffer size*/	
+		int buf_location =0; 	
+		do
+		{
+			current_rxbytes = recv(client_fd, (rx_buf+buf_location), BUFFER_SIZE-1, 0); 
+			if(current_rxbytes == -1) 
+			{
+				//perror("recv");
+				syslog(LOG_ERR, "sigprocmask");
+				close_all();
+				exit(EXIT_FAILURE);
+			} 
+			else 
+			{				
+				total_rxbytes += current_rxbytes;
+				buf_location  += current_rxbytes; 	
+			
+    			buffer_realloc_count++; //incrementing buffer size by * BUFFER_SIZE
+				rx_buf = (char*)realloc(rx_buf,(buffer_realloc_count)*(BUFFER_SIZE)*(sizeof(char)));				
+				//perror("realloc");
+				if(rx_buf == NULL)
+				{
+					syslog(LOG_ERR, "Error while reallocating buffer");
+					close_all();
+					exit(EXIT_FAILURE);
+				}
+
+			}
+		}while(strchr(rx_buf, '\n') == NULL);//Detect newline character - if ewline exists, break out of loop 
+		rx_buf[buf_location] = '\0';
+		
+		
+		/* Mask off signals while transferring data */
+		rc = sigprocmask(SIG_BLOCK, &mask, NULL);
+		if(rc!=0) 
+		{
+			perror("adding mask for signal");
+			syslog(LOG_ERR, "sigprocmask");
+			free(rx_buf);			 
+			close_all();
+		}
+		/* write full contents of buffer to ouputfile */
+		int written_bytes= write(outputfile_fd, rx_buf, buf_location); // Write to file
+		//perror("write");
+		if(written_bytes ==-1)
+		{
+			syslog(LOG_ERR, "Error while writing to o/p file");			
+			close_all();
+			exit(EXIT_FAILURE);
+		}
+		
+		// unmask signals after transfer finished  - remove singnal set from mask set 
+		rc = sigprocmask(SIG_UNBLOCK, &mask, NULL);	
+		//perror("sigprocmask unmasking");
+		if(rc!=0) 
+		{
+			syslog(LOG_ERR, "sigprocmask");
+			free(rx_buf);
+			close_all();		
+			exit(EXIT_FAILURE);
+		}
+		
+		
+		/* now copy full of file content into tx_buffer and send onto socket */		
+		
+		tx_buf = (char *)malloc(sizeof(char) * total_rxbytes);	  
```

Comment:
@MehulCUB what assumption is made here about the total size of the RAM heap relative to the total size of all packets sent?